core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
-            .map_err(anyhow::Error::from)
-            .typ(SystemError::Panic)?;
-        poller
-            .add(event_file.as_raw_fd(), Event::readable(42))
-            .map_err(anyhow::Error::from)
-            .typ(SystemError::Panic)?;
-
-        // We cannot use timeout directly and pass it to poller, as it may report a change of the
-        // event file that is unrelated to the populated property.
-        let mut leftover_time = timeout.saturating_sub(start.elapsed());
-        loop {
-            // Stop loop if time is exceeded
-            if leftover_time <= Duration::ZERO {
-                return Ok(false);
-            };
-
-            // Check if cgroup is populated (check may take time)
-            if !self.is_populated()? {
-                return Ok(true);
-            }
+    /// Kills all processes in this cgroup
+    pub fn kill(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
 
-            // Check again if time is exceeded
-            leftover_time = timeout.saturating_sub(start.elapsed());
-            if leftover_time <= Duration::ZERO {
-                return Ok(false);
-            };
-
-            // Poll wait for event with given leftover time as timeout
-            poller
-                .wait(Vec::new().as_mut(), Some(leftover_time))
-                .typ(SystemError::Panic)?;
-            poller
-                .modify(event_file.as_raw_fd(), Event::readable(42))
-                .map_err(anyhow::Error::from)
-                .typ(SystemError::Panic)?;
-
-            // determine new leftover time
-            leftover_time = timeout.saturating_sub(start.elapsed());
+        // We need to check for the existance of cgroup.kill, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.kill");
+        if !path.exists() {
+            bail!("cannot kill the root cgroup");
         }
+
+        Ok(fs::write(path, "1")?)
     }
 
-    /// Deletes a cgroup by killing it's processes and removing the directory afterwards
-    pub fn delete(&self, timeout: Duration) -> TypedResult<()> {
-        self.kill_all_timeout(timeout)?;
-        for d in WalkDir::new(&self.path)
-            .into_iter()
-            .flatten()
-            .filter(|e| e.file_type().is_dir())
-            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
-        {
-            std::fs::remove_dir(&d.path()).typ(SystemError::CGroup)?;
-            trace!("Removed {:?}", d.path().as_os_str())
+    /// Returns the path of this cgroup
+    pub fn get_path(&self) -> PathBuf {
+        self.path.clone()
+    }
+
+    /// Returns the path of the event file, which may be polled
+    pub fn get_events_path(&self) -> PathBuf {
+        self.path.join("cgroup.events")
+    }
+
+    /// Kills all processes and removes the current cgroup
+    pub fn rm(&self) -> anyhow::Result<()> {
```

Comment:
This function should probably consume the `CGroup` struct:  
`pub fn rm(self) -> anyhow::Result<()> {`

If we do this, we need to think about what to do in the case of this function returning an error.